### ISSUE TITLE
Universal Render Target

### ISSUE THREAD

**Issue by ryansolid** [solidjs/solid#593]
Solid's Babel Compiler can generate `dom` and `ssr` output. The `dom` output generates optimized template cloning and hydrates by walking over DOM elements. The `ssr` output generates string arrays that are built upon execution.

We also have a runtime for each. We use package.json export conditions to automatically alias based on the environment which runtime gets pulled. The babel plugin allows you to specify your runtime to import from. In general, we default that to `solid-js/web` but it could be anything.

Technically the Component creation code is shared between these as Components in Solid are just functions. If you never use a native element (lower case) and don't call render we never need to import DOM APIs. You can build whole systems with Solid just using components and the reactive system. Like here is Solid rendering WebGL: https://codesandbox.io/s/solid-gl-boxes-lmpei.

But you need to do a decent amount of manual wiring. Not a problem just a consideration. And while tree-shaking should drop unneeded DOM dependencies. But eventually, I imagine the desire to do more. 

So the plan here is to in addition to the other 2 provide a universal output. It might not be as optimized for the DOM or Server specifically, ie no cloning. But it would still group updates and handle reactive heuristics. The expectation is that anyone could implement a custom runtime to handle this output. What that runtime is could be anything.. WebGL, mobile, terminal whatever. All you need to do is implement the functions and publish an npm package. This package can also expose types via jsx-transform entry so that can be assigned to `jsxImportSource` in tsconfig. And then pass it to the `moduleName` option of `babel-preset-solid` and we're in business.

**Comment by vaukalak** [comment]
For the mobile output, I'd suggest to take to look to the [reanimated](https://docs.swmansion.com/react-native-reanimated/docs/animations) project, as a source for some thoughts. The main idea, is that this library creates a set of commands that run directly on the native thread of react-native, which avoids a lot of js<->native communication, and allow the app to be very snappy. There's also a youtube blog [can it be done with react-native](https://www.youtube.com/user/wcandill) which provides plenty of brilliant examples on how to use it. I would be cool, if the mobile adapter might be initially built with those optimizations techniques in mind.

**Comment by ryansolid** [comment]
Yeah I imagine a lot to be done here as soon as we open up the floodgates. Given it's all reactivity I think it will probably transfer to anything given you can attack it piecewise. That being said I don't know much about optimizing for mobile. If less communication is key here it might require a specialized transformation. The thing is with granular reactivity it's all about many small updates which sounds like communication might be expensive. The thinking here is it would be possible to apply granular updates to anything but if that doesn't work for certain cases those certain cases might need to be re-thought.

**Comment by joshuaellis** [comment]
> Like here is Solid rendering WebGL: https://codesandbox.io/s/solid-gl-boxes-lmpei.

This is a really cool thing to see. I suppose one downfall however is that you've had to manually call the classes within the component functions. When you're discussing a universal render target, would this open up the option to create a target for something like `threejs`. Similar to `react-three-fiber` which uses the `react-reconciler` to generate the classes based on elements i.e `<mesh />` internally calls `new THREE.Mesh` etc. 

If this were possible for solid i'd be incredibly interested in creating a port of `r3f` to solid. Maybe it's possible now and I haven't see the right way to do it?

**Comment by joshuaellis** [comment]
Hi @ryansolid thanks for the codesandbox. The renderer looks very interesting to see. I'll feedback here when I have an informed opinion.

**Comment by joshuaellis** [comment]
Hey @ryansolid 

I've had a chance to play around with the universal reconciler. And one piece of feedback // feature request would be to access the `props` passed to the node at time of creating (inside `createElement`). Why? Because there are classes that require arguments to be passed to their constructors in `three.js` e.g `InstancedMesh` & most geometries. To apply the props afterwards isn't effective and would probably include destroying the created instance and re-making it which seems unnecessary.

For context, here's my `createElement` function:

```js
createElement(el) {
      const constructorName = `${el.charAt(0).toUpperCase()}${el.slice(1)}`

      const target = THREE[constructorName]

      const instance = new target()

      if (constructorName.endsWith('Geometry')) {
        instance.s3f = {
          attach: 'geometry',
        }
      } else if (constructorName.endsWith('Material')) {
        instance.s3f = {
          attach: 'material',
        }
      }

      return instance
    }
```

**Comment by ryansolid** [comment]
Yeah this is what I was debating with the most. What you say makes sense. See React does everything create style, and Vue does it the way we went with. The problem of course is that React just calls createElement everytime, and we actually have fine-grained prop updates so there is an inconsistency. Also effects needs to run once so even if could pass props to `createElement` it would need to be only the non-dynamic ones, I think. Trying to think of ways we could have everything available at initial create without double executing all the dynamic attributes and still have tracking work properly.

I think the other piece is considerations around duplicate logic on the end user side having to implement both create and setProperty.. I have an idea.. what if we had an `endCreateElement` which we passed the element back to when we know we are done setting all the properties the first time.

```
const el = createElement("mesh");
setProperty(el, "someProp", someValue);
setProperty(el, "someProp2", someValue2);

// insert children
insert(el, createComponent(SomeComponent, { someComponentProp: someComponentValue })

createRenderEffect(() => {
  // diff and set dynamic props
  setProperty(el, "someProp3", someValue3);
})
endCreateElement(el); // called after first render
return el;
```
In this way it would give a potential point to collect initial things before doing some operation with it. But I suppose depending on the renderer separate granular updates don't help anyway, like when that effect re-runs maybe it updates 3 of 5 props it will call setProperty 3 times. Recreating instances seems awkward in this sense. [Vue](https://vuejs-course.com/blog/a-custom-renderer-for-vue-3) took this approach but maybe there is more to it. Maybe to use it this way renderer has to consider different type of batching? I'm not sure. I'd expect Vue to have the same concerns mentioned here.

This is the way Solid's DOM reconciler works, and using Components instead of native elements actually puts the reactivity on the implementor but seems better for this. I went that way first but the feedback was people were looking for a more traditional reconciler than just do whatever with the Components.

So not sure how we get granular updates without the implementor worrying about reactivity so that reconciliation can abstracted. One option would be `setProperties` being like an object instead of singular setter. Still getting the initial create to work the intended way would take some figuring out.
